core/core-impl/src/main/java/ai/timefold/solver/core/impl/util/ConstantLambdaUtils.java

```diff
@@ -0,0 +1,216 @@
+package ai.timefold.solver.core.impl.util;
+
+import java.util.Objects;
+import java.util.function.BiFunction;
+import java.util.function.BiPredicate;
+import java.util.function.Function;
+import java.util.function.ToIntBiFunction;
+import java.util.function.ToIntFunction;
+
+import ai.timefold.solver.core.api.function.QuadFunction;
+import ai.timefold.solver.core.api.function.ToIntQuadFunction;
+import ai.timefold.solver.core.api.function.ToIntTriFunction;
+import ai.timefold.solver.core.api.function.TriFunction;
+
+/**
+ * A class that hold common lambdas that are guarantee to be the same across method calls. In most JDK's, stateless lambdas are
+ * bound to a {@link java.lang.invoke.ConstantCallSite} inside the method that define them, but that
+ * {@link java.lang.invoke.ConstantCallSite} is not shared across methods (even for methods in the same class). Thus, when
+ * lambda reference equality is important (such as for node sharing in Constraint Streams), the lambdas in this class should be
+ * used.
+ */
+public final class ConstantLambdaUtils {
+    @SuppressWarnings("rawtypes")
+    private static final Function IDENTITY = Function.identity();
+
+    @SuppressWarnings("rawtypes")
+    private static final BiPredicate NOT_EQUALS = (a, b) -> !Objects.equals(a, b);
+
+    @SuppressWarnings("rawtypes")
+    private static final BiFunction BI_PICK_FIRST = (a, b) -> a;
+
+    @SuppressWarnings("rawtypes")
+    private static final TriFunction TRI_PICK_FIRST = (a, b, c) -> a;
+
+    @SuppressWarnings("rawtypes")
+    private static final QuadFunction QUAD_PICK_FIRST = (a, b, c, d) -> a;
+
+    @SuppressWarnings("rawtypes")
+    private static final BiFunction BI_PICK_SECOND = (a, b) -> b;
+
+    @SuppressWarnings("rawtypes")
+    private static final TriFunction TRI_PICK_SECOND = (a, b, c) -> b;
+
+    @SuppressWarnings("rawtypes")
+    private static final QuadFunction QUAD_PICK_SECOND = (a, b, c, d) -> b;
+
+    @SuppressWarnings("rawtypes")
+    private static final TriFunction TRI_PICK_THIRD = (a, b, c) -> c;
+
+    @SuppressWarnings("rawtypes")
+    private static final QuadFunction QUAD_PICK_THIRD = (a, b, c, d) -> c;
+
+    @SuppressWarnings("rawtypes")
+    private static final QuadFunction QUAD_PICK_FOURTH = (a, b, c, d) -> d;
+
+    @SuppressWarnings("rawtypes")
+    private static final ToIntFunction UNI_CONSTANT_ONE = (a) -> 1;
+    @SuppressWarnings("rawtypes")
+    private static final ToIntBiFunction BI_CONSTANT_ONE = (a, b) -> 1;
+    @SuppressWarnings("rawtypes")
+    private static final ToIntTriFunction TRI_CONSTANT_ONE = (a, b, c) -> 1;
+
+    @SuppressWarnings("rawtypes")
```

Comment:
Let's have consistent whitespace. Either use newlines, or don't.
If you want to separate blocks of lambdas, do so, but do it consistently.